Rust-native HackRf Support

TroyNeubauer opened the pull request:
Adds support for HackRF via a rust native driver powered by nusb.
- Rx works. Was able to run the FM demod example on android!
- TX is mostly implemented but untested

Closes #9

TroyNeubauer commented:
I am also working on a bladerf driver. This one is complicated enough that rewriting the c code in rust with nusb is impractical, so I plan on calling into `libbladerf` (no web support :cry: ).

What are the blocking / threading assumptions of `RxStreamer`? For this PR I assumed this was a blocking interface and used `futures_lite::block_on` to bridge the gap to async nusb. 

For the blade, this is a bit more complicated since `libbladerf` wraps libusb with their own interface: https://libusb.sourceforge.io/api-1.0/group__libusb__poll.html#ga6deff4c7d3a6c04bb9ec9fd259b48933, so I want to know a little more so I can plan out how to do this

bastibl reviewed (commented):
I get some errors, when I try to run the examples. When I run the `rx` example that is part of the hackrf crate, I get:
```
**basti@thrang git:(hackrf*):~/src/seify/crates/hackrf-one-rs >> cargo run --example rx
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.02s
     Running `/home/basti/.cargo/target/debug/examples/rx`
Board ID: Ok(2)
Version: Ok("2023.01.1")
Device version: UsbVersion(1, 0, 7)
Error: Failed to receive on hackrf

Caused by:
    0: transfer
    1: endpoint STALL condition**
```


When I try to run it with the seify example, I get:
```
basti@thrang git:(hackrf*):~/src/seify >> cargo run --example  probe --features=hackrfone
   Compiling seify v0.12.0 (/home/basti/src/seify)
error[E0433]: failed to resolve: use of undeclared crate or module `info`
  --> src/impls/hackrfone.rs:63:17
   |
63 | ...   info::warn!("HackRfOne::open received invalid args: bus_number: {bus_number:?}, ...
   |       ^^^^ use of undeclared crate or module `info`
```

TroyNeubauer commented on Cargo.toml:
```diff
@@ -24,6 +25,7 @@ futures = "0.3"
 log = "0.4"
 nom = "7.1"
 num-complex = "0.4"
+nusb = { git = "https://github.com/kevinmehall/nusb.git", optional = true }
```
Removed

TroyNeubauer commented on Cargo.toml:
```diff
@@ -32,6 +34,7 @@ thiserror = "1.0"
 [target.'cfg(not(target_arch = "wasm32"))'.dependencies]
 once_cell = "1.19"
 seify-rtlsdr = { path = "crates/rtl-sdr-rs", version = "0.0.3", optional = true }
+seify-hackrfone = { path = "crates/hackrf-one-rs", optional = true }
```
Sounds good, waiting on a new version from `nusb`: https://github.com/kevinmehall/nusb/issues/84, then will publish and update here

TroyNeubauer commented:
Splitting it into a separate github repo makes sense to me. Its now up at: https://github.com/MerchGuardian/seify-hackrfone, and I added a git submodule here.

I fixed the examples. I had to fix two funny issues:
1. I accidentally had `sample_rate_div` set to 0, which caused the stall since the device had no idea what to do lol.
2. Using `claim_interface` instead of `detach_and_claim_interface` would result in spurious `Device busy errors` on linux (worked fine without on MacOS)

I can now run both `seify-hackrfone` examples as well as `rx_generic` and `rx_threaded`!

bastibl commented:
Works for me too. Thank you! I'll merge and just revert the default features to keep soapy as the default.

bastibl commented:
I just tried to make a new Seify release, but that requires that your crate is published. Could you please do a `cargo publish`?

TroyNeubauer commented:
Just published `0.1.0`!
I disabled wrapping file descriptors for now while we wait on a new version of nusb: https://github.com/kevinmehall/nusb/issues/84
Once that is released, we can re-enable in `seify-hackrfone`, release, and bump in here

TroyNeubauer commented:
Made PR to bump here: https://github.com/FutureSDR/seify/pull/11. This also removes the `features = ["hackrfone"]` that I left by mistake
